# ASDF and World Coordinate Systems

The handling of World Coordinate Systems (WCS) is inextricably bound to 
digital observational data. For example, given a pixel in the data array 
from a detector, what RA, Dec does it map to? Or what pixel does an 
RA, Dec map to? Likewise for spectrographic data, what wavelength, 
RA, and Dec does a pixel correspond to? Astronomers have had various 
solutions to this problem in the past. Most of them have been awkward
to deal with.

Dealing with WCS issues flexibly was a major driver for ASDF and it 
deserves its own tutorial.

## FITS and WCS

FITS has had a series of papers and resulting standards to deal with
WCS issues. Generally speaking, these have presumed mostly issues of 
idealized detectors and optics, where distortion is not a significant
factor. There was much focus on handling many kinds of map projections
for images, but little provision for distortions beyond what projections
introduce. Likewise, most of the spectrosopic WCS models either presume
specific dispersion models, or the old fallback of wavelength tables,
which to this day many astronomers consider the only solution to 
spectroscopic WCS.

Handling distortions in FITS has met with a failure to settle on 
any general approach. Currently there are only two practical imaging
solutions, namely using the SIP convention, or a pixel translation
table. The first is not suitable for high spatial position precision
space telescopes 
such as HST, JWST or NGRST, the second is wasteful of disk space
while being inflexible to adjustments (more on that later). Since
most spectrographs use effectively the pixel translation approach,
they are also wasteful of space for IFUs and long slit data, while
also being inflexible to adjustments.

## A more flexible approach

Ideally, one would want to be able to construct coordinate transforms
out of simpler models to form a pipeline of transformations from 
one coordinate frame to another. The combination of the Python packages 
astropy modeling, GWCS (Generalized World Coordinate Systems) and asdf
provide a very flexible and powerful means of dealing with WCS.

This tutorial will not address how one develops a WCS model as that
is almost always a complex issue requiring very deep understanding
of the instrument involved and careful treatment of coordinates, and
usually is best left to experts (though being able to tweak the 
parameters of such WCS models is something than can and often should
be exposed to users).

This tutorial will focus more on using existing WCS models and seeing
how they can be modified to improve results.

## Using GWCS with ASDF

This initial imaging example will use the same wcs that was in the
JWST data file used in this tutorial: [Reading a JWST ASDF file]
(Reading_a_JWST_ASDF_file.ipynb) except only the WCS has been saved
to its own file (ASDF makes saving components of ASDF files to their
own files trivial).

## Required Software

This notebook requires that asdf 2.7.1 (or later), astropy 4.0.1 (or later) and gwcs 0.14 (or later) are installed

In [1]:
import asdf
from astropy.utils.data import download_file
import matplotlib
matplotlib.rcParams['figure.figsize'] = (10, 10)
from matplotlib import pyplot as plt
file1_url_1 = 'https://data.science.stsci.edu/redirect/ASDF/asdf_tut_gwcs.asdf'

In [2]:
filepath = download_file(file1_url_1)
# Wait until download completes (when 'In [*]' changes the  '* to a number')
# Ignore warnings

In [3]:
af = asdf.open(filepath)

c:\users\perry\documents\github\asdf\asdf\asdf.py:212: AsdfWarning: File 'file:///C:/Users/perry/AppData/Local/Temp/astropy-download-12004-c8lkqra6' was created with extension 'asdf.extension.BuiltinExtension' from package asdf-2.7.1, but older version asdf-2.6.1.dev59+gde25fec is installed
  warnings.warn(fmt_msg, AsdfWarning)


In [4]:
# Get the GWCS for these data
wcs = af['wcs']

In [5]:
# Illustrate use in converting pixel coordinate to sky
wcs(100,200)

[array(5.92172828), array(-72.0099005)]

In [6]:
# But what coordinate system for the outputs? What units?
wcs.output_frame

<CelestialFrame(name="world", unit=(Unit("deg"), Unit("deg")), axes_names=('lon', 'lat'), axes_order=(0, 1), reference_frame=<ICRS Frame>)>

In [7]:
# The WCS may have intermediate frames defined as well
wcs.pipeline

[Step(frame=detector,                       transform=CompoundModel),
 Step(frame=v2v3,                       transform=v23tosky),
 Step(frame=world,                       transform=None)]

From this we see that there are three frames defined, the detector frame
(e.g., pixel coordinates), the v2v3 frame, and the wolrd coordinate frame,
which is the output frame.

It is possible to access the corresponding coordinates for any intermediate frame.

In [8]:
v2v3coords = wcs.transform('detector','v2v3', 100, 200)
v2v3coords

[array(144.73670405), array(-545.14002711)]

In [9]:
# Units?
wcs.pipeline[1].frame.unit

(Unit("arcsec"), Unit("arcsec"))

In [10]:
# Complete computation to world
wcs.transform('v2v3', 'world', *v2v3coords)

(5.921728283943831, -72.00990050127562)

## Relevance to ASDF?

OK, GWCS has some extra features. But why is this part of an ASDF tutorial? Mainly to illustrate that ASDF preserves all the intrinsic information about the WCS and does not restrict transforms to a finite set of tools as FITS WCS does. We will illustrate this by examining this WCS in more detail. First note that ASDF allows an arbitrary number of intermediate frames (especially useful for spectral data), as well as an arbitrary number of different WCS models within a file. ASDF provides a way of saving most of astropy's models, of which there is a large number. More important, it can save astropy compound models. Arbitrarily complex combinations of models can be used and saved in ASDF, including all the projection models that FITS provides.

So while most of this tutorial is focused on GWCS, it is ASDF that enables this to be used with data files without forcing the user to obtain reference files and special software to determine the correct WCS as had to be done during most of HST lifetime.

Let's look at the details of the first step's transform.

In [11]:
transform1 = wcs.pipeline[0].transform
type(transform1)

<class 'astropy.modeling.core.CompoundModel'>
Name: CompoundModel
N_inputs: <property object at 0x000002AB0700AB38>
N_outputs: <property object at 0x000002AB0700AB88>
Fittable parameters: <property object at 0x000002AB0700AA98>

In [12]:
print(transform1)

Model: CompoundModel
Inputs: ('x0', 'x1')
Outputs: ('y0', 'y1')
Model set size: 1
Expression: [0] & [1] | [2] & [3] | [4] | [5] & [6] | [7] | [8] & [9] | [10] & [11]
Components: 
    [0]: <Shift(offset=1.)>

    [1]: <Shift(offset=1.)>

    [2]: <Shift(offset=-1024.5)>

    [3]: <Shift(offset=-1024.5)>

    [4]: <Mapping((0, 1, 0, 1))>

    [5]: <Polynomial2D(5, c0_0=-0., c1_0=0.06277675, c2_0=0.00000012, c3_0=0., c4_0=0., c5_0=0., c0_1=0., c0_2=-0.0000001, c0_3=-0., c0_4=-0., c0_5=0., c1_1=-0.00000072, c1_2=0., c1_3=-0., c1_4=0., c2_1=-0., c2_2=0., c2_3=0., c3_1=-0., c3_2=0., c4_1=0.)>

    [6]: <Polynomial2D(5, c0_0=0., c1_0=-0.00009103, c2_0=0.00000031, c3_0=-0., c4_0=0., c5_0=0., c0_1=0.06307627, c0_2=-0.00000042, c0_3=0., c0_4=-0., c0_5=0., c1_1=0.00000023, c1_2=-0., c1_3=0., c1_4=0., c2_1=0., c2_2=-0., c2_3=0., c3_1=0., c3_2=0., c4_1=0.)>

    [7]: <Mapping((0, 1, 0, 1))>

    [8]: <Polynomial2D(1, c0_0=0., c1_0=-0.99999917, c0_1=-0.00128895)>

    [9]: <Polynomial2D(1, c0_0=0., 

This compound model is sufficiently complicated to require some detailed explanation (but not nearly as complicated as some JWST spectral WCS models!). First a brief explanation of the notation. Inputs and outputs indicate that this model takes two input values (or two arrays of values and generates two output values, i.e., two input coordinates and two output coordinates. Expression is the expression used to combine astropy models into a more complex compound model. To keep the expression concise, each term is replaced by a number and the model associated with that number is listed below. The two operators in particular that need special explanation are & and |. The & operator means that two functions are joined by adding dimensionality. If a Shift function shifts the value of the input by the given value, the joining two Shift functions shift the first input by the first value, and the second input by the second value. It is a way of making two 1D functions a 2D function, albeit with each input coordinate treated independently of the other. The | operator feeds the output of the preceding function into the next. The number of outputs of the preceding must match the number of inputs of the following function. Finally, the Mapping function acts as a switchboard between inputs and output. For example, Mapping(0, 1, 0, 1) takes two inputs and generates 4 outputs, where outputs 0, and 2 are both equal to the first input, and outputs 1 and 3 are both equal to the second input.

Now to summarize in words what this transform does: shifts both input values first by 1 and then by -1024.5. This to make the coordinates relative to the center of the detector and center of that pixel. The next step splits the two input values so that both go to two different 2D polynomials (each of which produces only one value, so the net result is two output values. Then these values go through a similar split and to a second polynomial that is effectively applying a small rotation, followed by a shift in the V2,V3 coordinate system.

Now to look at the second step transform.

In [13]:
transform2 = wcs.pipeline[1].transform
print(transform2)

Model: CompoundModel
Name: v23tosky
Inputs: ('x0', 'x1')
Outputs: ('lon', 'lat')
Model set size: 1
Expression: [0] & [1] | [2] | [3] | [4]
Components: 
    [0]: <Scale(factor=0.00027778)>

    [1]: <Scale(factor=0.00027778)>

    [2]: <SphericalToCartesian()>

    [3]: <RotationSequence3D(angles=[  0.02391763,   0.13700764, 359.92586311, -71.99550858,  -5.86893417])>

    [4]: <CartesianToSpherical()>
Parameters:
           factor_0       ...                angles_3 [5]               
    --------------------- ... ------------------------------------------
    0.0002777777777777778 ... 0.023917627222222224 .. -5.868934166666667


This transform is much simpler and consists of a scale change then conversion to cartesian to perform Euler rotations to the RA, Dec system and then conversion back to spherical coordinates.

Note that there is no use of a tangent projection, and such distortions that would result from that have been folded into the polynomial distortion model.

## Modifying the WCS

This is a somewhat contrived example, but it is to illustrate that one can make customizations to a WCS fairly easily. In this particular case one could change any of the parameters in the transforms, and save the new WCS to an ASDF file without any concern for how to do that. Or one can modify the existing transform by adding new elements to the transform. So supposing the existing plate scale used does not include velocity aberration, then one would want to include that as well as a plate scale change, and perhaps not change  the existing plate scale application to be clear there are two different scales applied. 



In [14]:
from astropy.modeling.models import Scale
va_plate_scale_correction = 1.00008
transform3 = (Scale(va_plate_scale_correction) & Scale(va_plate_scale_correction)) | transform2
wcs.pipeline[1].transform = transform3
af.write_to('modified_wcs.asdf', all_array_storage='inline')

In [15]:
af2 = asdf.open('modified_wcs.asdf')

In [16]:
wcs2 = af2['wcs']
mtransform2 = wcs2.pipeline[1].transform
print(mtransform2)

Model: CompoundModel
Inputs: ('x0', 'x1')
Outputs: ('lon', 'lat')
Model set size: 1
Expression: [0] & [1] | [2] & [3] | [4] | [5] | [6]
Components: 
    [0]: <Scale(factor=1.00008)>

    [1]: <Scale(factor=1.00008)>

    [2]: <Scale(factor=0.00027778)>

    [3]: <Scale(factor=0.00027778)>

    [4]: <SphericalToCartesian()>

    [5]: <RotationSequence3D(angles=[  0.02391763,   0.13700764, 359.92586311, -71.99550858,  -5.86893417])>

    [6]: <CartesianToSpherical()>
Parameters:
    factor_0 factor_1 ...                angles_5 [5]               
    -------- -------- ... ------------------------------------------
     1.00008  1.00008 ... 0.023917627222222224 .. -5.868934166666667


For those interested, here are the actual contents of the asdf file

In [17]:
with open('modified_wcs.asdf') as asdftext: print(asdftext.read()) 

#ASDF 1.0.0
#ASDF_STANDARD 1.5.0
%YAML 1.1
%TAG ! tag:stsci.edu:asdf/
--- !core/asdf-1.1.0
asdf_library: !core/software-1.0.0 {author: Space Telescope Science Institute, homepage: 'http://github.com/spacetelescope/asdf',
  name: asdf, version: 2.6.1.dev59+gde25fec}
history:
  extensions:
  - !core/extension_metadata-1.0.0
    extension_class: asdf.extension.BuiltinExtension
    software: !core/software-1.0.0 {name: asdf, version: 2.6.1.dev59+gde25fec}
  - !core/extension_metadata-1.0.0
    extension_class: gwcs.extension.GWCSExtension
    software: !core/software-1.0.0 {name: gwcs, version: 0.14.0}
wcs: !<tag:stsci.edu:gwcs/wcs-1.1.0>
  name: ''
  steps:
  - !<tag:stsci.edu:gwcs/step-1.1.0>
    frame: !<tag:stsci.edu:gwcs/frame2d-1.0.0>
      axes_names: [x, y]
      axes_order: [0, 1]
      axis_physical_types: ['custom:x', 'custom:y']
      name: detector
      unit: [!unit/unit-1.0.0 pixel, !unit/unit-1.0.0 pixel]
    transform: !transform/compose-1.2.0
      bounding_box:
      - [

The result is fairly extensive and not nearly as easy to interpret as the python printed version. Nevertheless, it demonstrates ASDF is capable of saving arbitrarily complex expressions without the user needing to understand the details of saving or reading those expressions.